external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/DirLock.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.storm.hdfs.spout;
+
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hdfs.DistributedFileSystem;
+import org.apache.storm.hdfs.common.HdfsUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Facility to synchronize access to HDFS directory. The lock itself is represented
+ * as a file in the same directory. Relies on atomic file creation.
+ */
+public class DirLock {
+  private FileSystem fs;
+  private final Path lockFile;
+  public static final String DIR_LOCK_FILE = "DIRLOCK";
+  private static final Logger LOG = LoggerFactory.getLogger(DirLock.class);
+  private DirLock(FileSystem fs, Path lockFile) throws IOException {
+    if( fs.isDirectory(lockFile) ) {
+      throw new IllegalArgumentException(lockFile.toString() + " is not a directory");
+    }
+    this.fs = fs;
+    this.lockFile = lockFile;
+  }
+
+  /** Get a lock on file if not already locked
+   *
+   * @param fs
+   * @param dir  the dir on which to get a lock
+   * @return The lock object if it the lock was acquired. Returns null if the dir is already locked.
+   * @throws IOException if there were errors
+   */
+  public static DirLock tryLock(FileSystem fs, Path dir) throws IOException {
+    Path lockFile = getDirLockFile(dir);
+
+    try {
+      FSDataOutputStream ostream = HdfsUtils.tryCreateFile(fs, lockFile);
```

Comment:
You are correct local file system does not do anything atomic.  Even rename is not atomic, even though the spec says it should be.  we should not test with the local file system if we want things to be atomic.  I think we mostly want to call out in the docs that if it was not obvious already the FileSystem used needs to support a specific set of atomic operations. (MRv2 has similar constraints for AM recovery, but it is not documented very well)